[mbed-os-5.15] Network interface MAC address set/get

artokin opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
Add method set_mac_address that Application can use to set MAC address to the interface.
Update get_mac_address to return MAC address correctly from EMAC.

This PR is copied from: https://github.com/ARMmbed/mbed-os/pull/13902


#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@juhhei01 , @JarkkoPaso , @mikaleppanen , @mikter 

----------------------------------------------------------------------------------------------------------------

Commit:
Method for adding network interface MAC address

Add method set_mac_address to set network interface MAC address.

Commit:
MAC address set/get support for EMAC interface

Add support to set and get MAC address from EMAC interface.

Commit:
Review correction: Remove unnecessary method

Remove unnecessary method: add_ethernet_interface_ns

0xc0170 reviewed (approved):
No tests

Commit:
Update unittests

Commit:
Update set_mac_address signature

Use nsapi_size_t instead of size_t.
Update missing method description.

mikter commented on features/netsocket/NetworkInterface.h:
```diff
@@ -100,6 +100,24 @@ class NetworkInterface: public DNS {
+    /** Set the MAC address to the interface.
+     *
+     *  Provided MAC address is set the the network interface. Address must be
+     *  set before calling the interface connect() method. EUI-48 MAC addresses
```
My only question here is should we mention that this must be unique globally and there are places where to get these and also that some interfaces already have an unique MAC address written in HW and this should not be called on those interfaces

Sort of own chapter on the possible hazards if this API is called without proper knowledge

artokin replied:
Method description updated, please re-review.

Commit:
Review corrections

-Update method set_mac_address description
-Update EMAC test method return value

Commit:
Update EMAC connect logic with MAC address change

Do not fail EMAC interface during connect if MAC address change is
not supported in the selected interface.

Commit:
Follow Mbed OS coding style

Commit:
Fix method description

Commit:
Fix unittests

Use path in "mbed-trace/mbed_trace.h" to keep unittest happy.

Commit:
Update EmacTestNetworkStack

-Uncomment get_mac_address
-Document set_mac_address

Commit:
Remove API changes

-Remove Nanostack::add_ethernet_interface API change
-Add get_mac_address to MeshEthernetInterface

Commit:
Update WICED binaries for ARMC6

A new method has been added to the NetworkInterface. Therfore WICED
library needs to be rebuild for ARMC6.

Update WICED binaries for:
-MTB_ADV_WISE_1530
-MTB_MXCHIP_EMW3166
-MTB_USI_WM_BN_BM_22

andypowers reviewed (approved):
Approved.
